addOns/authhelper/src/main/java/org/zaproxy/addon/authhelper/AuthUtils.java

```diff
@@ -1164,6 +1173,126 @@ public static boolean isRelevantToAuth(HttpMessage msg) {
+     * is returned in HTML. The method first tries the URL without a path before falling back to the
+     * given URL.
+     *
+     * @param authSender
+     * @param user
+     * @param session
+     * @param loginUrl
+     */
+    public static void checkLoginLinkVerification(
+            HttpSender authSender, User user, WebSession session, String loginUrl) {
+        AuthCheckingStrategy verif =
+                user.getContext().getAuthenticationMethod().getAuthCheckingStrategy();
+        if (!AuthCheckingStrategy.AUTO_DETECT.equals(verif)) {
+            return;
+        }
+        try {
+            URI testUri = new URI(loginUrl, true);
+            if (StringUtils.isEmpty(testUri.getPath())) {
+                // Try to top level link first, if the page has the login form then its less likely
+                // to have a link to one
+                testUri.setPath("");
+                if (checkLoginLinkVerification(authSender, user, session, testUri)) {
+                    // The top level URL worked :)
+                    return;
+                }
+                testUri = new URI(loginUrl, true);
+            }
+            checkLoginLinkVerification(authSender, user, session, testUri);
+
+        } catch (Exception e) {
+            LOGGER.warn(
+                    "Failed accessing potential login link verification URL {}, {}",
+                    loginUrl,
+                    e.getMessage(),
+                    e);
+        }
+    }
+
+    private static boolean checkLoginLinkVerification(
+            HttpSender authSender, User user, WebSession session, URI testUri) {
+        try {
+            // Send an unauthenticated req to the test site, manually following redirects as needed
+            HttpMessage msg = new HttpMessage(testUri);
+            HttpSender unauthSender = new HttpSender(HttpSender.AUTHENTICATION_HELPER_INITIATOR);
+            unauthSender.sendAndReceive(msg);
+            AuthenticationHelper.addAuthMessageToHistory(msg);
+            int count = 0;
+            while (HttpStatusCode.isRedirection(msg.getResponseHeader().getStatusCode())) {
+                testUri =
+                        new URI(
+                                msg.getResponseHeader().getHeader(HttpResponseHeader.LOCATION),
+                                true);
+                msg = new HttpMessage(testUri);
+                unauthSender.sendAndReceive(msg);
+                AuthenticationHelper.addAuthMessageToHistory(msg);
+                if (count++ > 50) {
+                    return false;
+                }
+            }
+
+            if (!msg.getResponseHeader().isHtml()) {
+                LOGGER.debug(
+                        "Response to {} is no good as a login link verification req, it is not HTML {}",
+                        testUri,
+                        msg.getResponseHeader().getNormalisedContentTypeValue());
```

Comment:
Is HTML too strict? Or is that a question for another day.

I'm thinking some may use an api which simply return json or xml 🤷‍♂️

Reply:
Just HTML for now as we'll only detect links in HTML 😁 I would like to extend this to json / xml, but then how to detect login links?

Reply:
Okay, will think about it